snapcraft/internal/meta/_snap_packaging.py

```diff
@@ -133,6 +133,97 @@ def _update_yaml_with_extracted_metadata(
         raise meta_errors.MissingSnapcraftYamlKeysError(keys=missing_keys)
 
 
+def _adopt_info(
+        config_data: Dict[str, Any],
+        extracted_metadata: _metadata.ExtractedMetadata):
+    metadata_dict = extracted_metadata.to_dict()
+    for key, value in filter(
+            lambda item: item != 'desktop',
+            metadata_dict.items()):
+        if key not in config_data:
+            if key == 'icon':
+                if _icon_file_exists() or not os.path.exists(value):
+                    # Do not overwrite the icon file.
+                    continue
+            config_data[key] = value
+    if 'desktop' in metadata_dict:
+        for desktop_id in metadata_dict['desktop']:
+            app_name = _get_app_name_from_desktop_id(config_data, desktop_id)
+            if app_name and not _desktop_file_exists(app_name):
```

Comment:
If not app_name, would it be helpful to display a warning?

Reply:
Maybe. This seemed to me like a best-effort extractor, trying to match as much as possible but not complaining too much, so I didn't put any warning. But we certainly can.
However, because the appstream file can have multiple desktop files, there will be cases in which the snap only defines a few of the apps that have a launchable in appstream. Do you think we should warn in such cases?

Reply:
Ah yes. No, I think what you've done is correct.